### PR TITLE
feat:(@desktop/channels): loading state when switching channels/chats

### PR DESCRIPTION
**Pull request by mprakhov**
### What does the PR do

- show loading state when switching channels/chats until initial messages loading
- make `getFirstUnseenMessageIdFor` call async

Closes: #9441

### Affected areas

1. Showing chat content during switching between the channels
2. First unseen message marker

### Screenshot of functionality (including design for comparison)

https://user-images.githubusercontent.com/117639195/227528514-222e4bcb-17f5-4a02-b84b-a459cb7d9833.mov

### CODE REVIEW

File: src/app_service/service/message/service.nim [hunk 1]
```diff
@@ -188,9 +193,6 @@ QtObject:
     let pinnedMsgCursorValue = if (pinnedMsgCursor.isFetchable()): pinnedMsgCursor.getValue() else: CURSOR_VALUE_IGNORE
 
     if(msgCursorValue == CURSOR_VALUE_IGNORE and pinnedMsgCursorValue == CURSOR_VALUE_IGNORE):
-      # it's important to emit signal in case we are not fetching messages, so we can update the view appropriatelly.
-      let data = MessagesLoadedArgs(chatId: chatId)
-      self.events.emit(SIGNAL_MESSAGES_LOADED, data)
```
Comment (jrainville) [inline src/app_service/service/message/service.nim:193]: Why did you remove this?

**Review by jrainville — COMMENTED** [review]
Nice job. It's way cleaner.
There are still some stutters, but it's already way better. We can fix those later.

File: src/app_service/service/message/service.nim (hunk 1 above)
Comment (mprakhov) [reply src/app_service/service/message/service.nim:193]: It looks like some legacy part of the code which does not need now.
When we open a new chat, we receive a few `SIGNAL_MESSAGES_LOADED` events, this one the first and the second event - with the initial messages.
SIGNAL_MESSAGES_LOADED event subscriber mark `firstUnseenMessageLoaded` as true, but in reality, we didn't get the messages. So we started displaying chat UI while messages are still loading and we can observe behavior, when we have chat name/icon, and fetch content, but after a few seconds, messages are loading.

I see, that this event emitted again from this place only when we scrolled to the top of available messages.
I did not find what it breaks

**Review by osmaczko — DISMISSED** [review]
Very nice!

One behavior is broken though, it is described in the video:

https://user-images.githubusercontent.com/33099791/227554182-95519755-9628-41f6-96cf-2264e2898eee.mp4

File: src/app_service/service/message/service.nim (hunk 1 above)
Comment (jrainville) [reply src/app_service/service/message/service.nim:193]: I **think** it used to make it hide the little loader at the top of a chat when we start fetching messages. Do you see that it stays forever now, especially when you scroll to the top of a channel?

Comment (mprakhov) [reply src/app_service/service/message/service.nim:193]: based on what I see, it is just stopping processing new messages fetch when we do not need it.
It sends empty data and doesn't impact the UI (at least I did not find).
SIGNAL_MESSAGES_LOADED will be sent second time, but with messages (or without, if there are no messages)

**Comment by mprakhov** [comment]
@osmaczko fixed

https://user-images.githubusercontent.com/117639195/227578683-bbfdab60-7410-4dc4-b387-e42c30fcee3d.mov

**Review by jrainville — APPROVED** [review]

File: src/app/modules/main/chat_section/chat_content/messages/module.nim [hunk 2]
```diff
@@ -721,3 +717,11 @@ proc updateItemsByAlbum(self: Module, items: var seq[Item], message: MessageDto)
         items[i] = item
         return true
   return false
+
+method onFirstUnseenMessageId*(self: Module, messageId: string) =
+  var isChatOpened = self.view.getFirstUnseenMessageLoaded()
```
[the remark this answers is not in the document]
Comment (osmaczko) [reply src/app/modules/main/chat_section/chat_content/messages/module.nim:722]: It doesn't work as expected in the following scenario (it should scroll to `----` message):


https://user-images.githubusercontent.com/33099791/227888463-04fa4295-ab36-415f-844d-206a28b508ce.mp4

**Review by osmaczko — CHANGES_REQUESTED** [review]

File: src/app/modules/main/chat_section/chat_content/messages/module.nim (hunk 2 above)
[the remark this answers is not in the document]
Comment (mprakhov) [reply src/app/modules/main/chat_section/chat_content/messages/module.nim:722]: @osmaczko fixed

File: ui/app/AppLayouts/Chat/views/ChatMessagesView.qml [hunk 3]
```diff
@@ -83,6 +83,13 @@ Item {
             chatLogView.itemAtIndex(messageIndex).startMessageFoundAnimation()
         }
 
+        function onScrollToFirstUnreedMessage(messageIndex) {
```
Comment (osmaczko) [inline ui/app/AppLayouts/Chat/views/ChatMessagesView.qml:86]:
```suggestion
        function onScrollToFirstUnreadMessage(messageIndex) {
```

File: src/app/modules/main/chat_section/chat_content/messages/view.nim [hunk 4]
```diff
@@ -233,4 +234,19 @@ QtObject:
...
+    self.firstUnseenMessageLoaded = value
+    self.firstUnseenMessageLoadedChanged()
+  
+  QtProperty[bool] firstUnseenMessageLoaded:
+    read = getFirstUnseenMessageLoaded
+    notify = firstUnseenMessageLoadedChanged
+
+  proc scrollToFirstUnreedMessage(self: View, messageIndex: int) {.signal.}
+  proc emitScrollToFirstUnreedMessageSignal*(self: View, messageIndex: int) =
+    self.scrollToFirstUnreedMessage(messageIndex)
```
Comment (osmaczko) [inline src/app/modules/main/chat_section/chat_content/messages/view.nim:252]: unreed -> unread
```suggestion
  proc scrollToFirstUnreadMessage(self: View, messageIndex: int) {.signal.}
  proc emitScrollToFirstUnreadMessageSignal*(self: View, messageIndex: int) =
    self.scrollToFirstUnreadMessage(messageIndex)
```

**Review by osmaczko — APPROVED** [review]

File: src/app/modules/main/chat_section/chat_content/messages/controller.nim [hunk 5]
```diff
@@ -280,7 +286,7 @@ proc setSearchedMessageId*(self: Controller, searchedMessageId: string) =
 proc clearSearchedMessageId*(self: Controller) =
   self.setSearchedMessageId("")
 
-proc getFirstUnseenMessageId*(self: Controller): string =
+proc getFirstUnseenMessageId*(self: Controller) =
```
[the remark this answers is not in the document]
Comment (osmaczko) [reply src/app/modules/main/chat_section/chat_content/messages/controller.nim:289]: Would it make sense to change the name to `fetchFirstUnseenMessageId` or `getAsyncFirstUnseenMessageId` ?
